lib/PhasarLLVM/TypeHierarchy/DIBasedTypeHierarchy.cpp

```diff
@@ -0,0 +1,272 @@
+/******************************************************************************
+ * Copyright (c) 2023 Fabian Schiebel.
+ * All rights reserved. This program and the accompanying materials are made
+ * available under the terms of LICENSE.txt.
+ *
+ * Contributors:
+ *     Fabian Schiebel and others
+ *****************************************************************************/
+
+#include "phasar/PhasarLLVM/TypeHierarchy/DIBasedTypeHierarchy.h"
+
+#include "phasar/PhasarLLVM/DB/LLVMProjectIRDB.h"
+#include "phasar/PhasarLLVM/TypeHierarchy/LLVMVFTable.h"
+#include "phasar/TypeHierarchy/VFTable.h"
+
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/BinaryFormat/Dwarf.h"
+#include "llvm/IR/DebugInfoMetadata.h"
+#include "llvm/IR/DerivedTypes.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/GlobalVariable.h"
+#include "llvm/IR/Metadata.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/ErrorHandling.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include <cassert>
+#include <iostream>
+
+namespace psr {
+
+DIBasedTypeHierarchy::DIBasedTypeHierarchy(const LLVMProjectIRDB &IRDB) {
+  const auto *const Module = IRDB.getModule();
+  Finder.processModule(*Module);
+
+  // find and save all base types first, so they are present in TypeToVertex
+  size_t Counter = 0;
+  for (const llvm::DIType *DIType : Finder.types()) {
+    if (const auto *CompositeType =
+            llvm::dyn_cast<llvm::DICompositeType>(DIType)) {
+      NameToType.try_emplace(CompositeType->getName(), CompositeType);
+      TypeToVertex.try_emplace(CompositeType, Counter++);
+      VertexTypes.push_back(CompositeType);
+      continue;
+    }
+  }
+
+  std::vector<llvm::SmallVector<size_t>> DerivedTypesOf;
+  // find and save all derived types
+  for (const llvm::DIType *DIType : Finder.types()) {
+    if (const auto *DerivedType = llvm::dyn_cast<llvm::DIDerivedType>(DIType)) {
+      if (DerivedType->getTag() == llvm::dwarf::DW_TAG_inheritance) {
+        const size_t ActualDerivedType =
+            TypeToVertex[NameToType[DerivedType->getScope()->getName()]];
+
+        // (max) assertions fail, but code works?!
+        // assert(ActualDerivedType);
+
+        size_t BaseTypeVertex = TypeToVertex[DerivedType->getBaseType()];
+        // (max) assertions fail, but code works?!
+        // assert(BaseTypeVertex);
+
+        llvm::SmallVector<size_t> BaseType = {
+            BaseTypeVertex, static_cast<size_t>(ActualDerivedType)};
+        DerivedTypesOf.push_back(BaseType);
+        continue;
+      }
+    }
+  }
+
+  // Initialize the transitive closure matrix with all as false
+  llvm::BitVector InitVector(VertexTypes.size(), false);
+
+  for (size_t I = 0; I < VertexTypes.size(); I++) {
+    TransitiveClosure.push_back(InitVector);
+  }
+
+  // set direct edges
+  for (const auto &Vertex : DerivedTypesOf) {
+    TransitiveClosure[Vertex[0]][Vertex[1]] = true;
+  }
```

Comment:
Do you stil need DerivedTypesOf? Seems you can set TransitiveClosure directly in the loop above